File: src/lib/gyftr.ts
import "server-only";
import crypto from "node:crypto";
import { env } from "@/lib/env";
import type { OrderItem } from "@/types";
import type { NewVoucher } from "@/lib/db/types";

/**
 * Client for the Gyftr core APIs (catalog + voucher issuance).
 *
 * When GYFTR_USE_MOCK is true (default until real credentials are supplied)
 * the catalog/issuance are simulated locally so the portal is fully
 * functional end-to-end. Flip GYFTR_USE_MOCK=false and set GYFTR_API_KEY to
 * route these calls to the real Gyftr gateway.
 */

function randomCode(brandExt: string, denomination: number): string {
  const block = () => crypto.randomBytes(2).toString("hex").toUpperCase();
  return `${brandExt}-${block()}-${block()}-${denomination}`;
}
function randomPin(): string {
  return crypto.randomInt(0, 1_000_000).toString().padStart(6, "0");
}

export interface IssueVoucherRequest {
  orderNumber: string;
  items: OrderItem[];
}

/** Issues vouchers for an order. Returns flat list ready to persist. */
export async function issueVouchers(req: IssueVoucherRequest): Promise<NewVoucher[]> {
  if (!env.gyftr.useMock) {
    return issueViaGyftrApi(req);
  }
  return issueMock(req);
}

function issueMock(req: IssueVoucherRequest): NewVoucher[] {
  const expiry = new Date();
  expiry.setFullYear(expiry.getFullYear() + 1);
  const expiryDate = expiry.toISOString().slice(0, 10);
  const out: NewVoucher[] = [];
  for (const item of req.items) {
    const ext = (item.brandName.replace(/[^A-Z]/gi, "").slice(0, 4) || "GYF").toUpperCase();
    for (let i = 0; i < item.quantity; i++) {
      out.push({
        orderItemId: item.id,
        brandId: item.brandId,
        brandName: item.brandName,
        denomination: item.denomination,
        code: randomCode(ext, item.denomination),
        pin: randomPin(),
        expiryDate,
      });
    }
  }
  return out;
}

async function issueViaGyftrApi(req: IssueVoucherRequest): Promise<NewVoucher[]> {
  const url = `${env.gyftr.baseUrl}${env.gyftr.sendVoucherEndpoint}`;
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${env.gyftr.apiKey}`,
    },
    body: JSON.stringify({
      orderNumber: req.orderNumber,
      items: req.items.map((i) => ({
        brand: i.brandId,
        denomination: i.denomination,
        quantity: i.quantity,
      })),
    }),
  });
  if (!res.ok) {
    throw new Error(`Gyftr voucher issuance failed: ${res.status} ${await res.text()}`);
  }
  const json = (await res.json()) as {
    vouchers: Array<{
      orderItemId?: string;
      brandId?: string;
      brandName: string;
      denomination: number;
      code: string;
      pin?: string;
      expiryDate?: string;
    }>;
  };
  return json.vouchers.map((v) => ({
    orderItemId: v.orderItemId ?? "",
    brandId: v.brandId ?? null,
    brandName: v.brandName,
    denomination: v.denomination,
    code: v.code,
    pin: v.pin ?? null,
    expiryDate: v.expiryDate ?? null,
  }));
}

File: core/services/gyftr.py
"""
Gyftr core API client (catalog + voucher issuance) — mirrors `src/lib/gyftr.ts`.

With GYFTR_USE_MOCK true (default) issuance is simulated locally so the portal
is fully functional end-to-end. Flip the flag + set GYFTR_API_KEY for real calls.
"""
import re
import secrets
from datetime import date

from django.conf import settings


def _random_code(brand_ext: str, denomination: int) -> str:
    block = lambda: secrets.token_hex(2).upper()
    return f"{brand_ext}-{block()}-{block()}-{denomination}"


def _random_pin() -> str:
    return str(secrets.randbelow(1_000_000)).zfill(6)


def issue_vouchers(order_number: str, items) -> list[dict]:
    """`items` is an iterable of OrderItem model instances. Returns NewVoucher dicts."""
    if not settings.APP["gyftr"]["use_mock"]:
        return _issue_via_api(order_number, items)
    return _issue_mock(items)


def _issue_mock(items) -> list[dict]:
    today = date.today()
    try:
        expiry = today.replace(year=today.year + 1)
    except ValueError:  # Feb 29
        expiry = today.replace(year=today.year + 1, day=28)
    expiry_date = expiry.isoformat()

    out: list[dict] = []
    for item in items:
        letters = re.sub(r"[^A-Za-z]", "", item.brand_name)[:4]
        ext = (letters or "GYF").upper()
        for _ in range(item.quantity):
            out.append({
                "orderItemId": item.id,
                "brandId": item.brand_id,
                "brandName": item.brand_name,
                "denomination": item.denomination,
                "code": _random_code(ext, item.denomination),
                "pin": _random_pin(),
                "expiryDate": expiry_date,
            })
    return out


def _issue_via_api(order_number: str, items) -> list[dict]:
    import requests

    g = settings.APP["gyftr"]
    url = f"{g['base_url']}{g['send_voucher_endpoint']}"
    res = requests.post(
        url,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {g['api_key']}",
        },
        json={
            "orderNumber": order_number,
            "items": [
                {"brand": i.brand_id, "denomination": i.denomination, "quantity": i.quantity}
                for i in items
            ],
        },
        timeout=30,
    )
    res.raise_for_status()
    payload = res.json()
    return [
        {
            "orderItemId": v.get("orderItemId", ""),
            "brandId": v.get("brandId"),
            "brandName": v["brandName"],
            "denomination": v["denomination"],
            "code": v["code"],
            "pin": v.get("pin"),
            "expiryDate": v.get("expiryDate"),
        }
        for v in payload.get("vouchers", [])
    ]
